client/minicomm/subprocess_delegator.cc

```diff
@@ -0,0 +1,298 @@
+      config_->SubprocessConfig();
+  if (config_proto.filename().empty()) {
+    LOG(ERROR) << "Subprocess not configured.";
+    return;
+  }
+  std::unique_lock<std::mutex> pid_lock(child_pid_mutex_);
+  // If there is already child, or we are shutting down, we do nothing.
+  if (child_pid_ != 0) {
+    return;
+  }
+  int stdin_fd[2];
+  int stdout_fd[2];
+  int stderr_fd[2];
+  pipe(stdin_fd);
+  pipe(stdout_fd);
+  pipe(stderr_fd);
+  child_pid_ = fork();
+  if (child_pid_ == 0) {
+    close(stdin_fd[1]);
+    close(stdout_fd[0]);
+    close(stderr_fd[0]);
+
+    vector<const char*> newargv;
+    newargv.reserve(config_proto.argv_size() + 2);
+    newargv.push_back(config_proto.filename().c_str());
+    for (const string& arg : config_proto.argv()) {
+      newargv.push_back(arg.c_str());
+    }
+    newargv.push_back(NULL);
+
+    vector<const char*> newenv;
+    newenv.reserve(config_proto.env_size() + 1);
+    for (const string& v : config_proto.env()) {
+      newenv.push_back(v.c_str());
+    }
+    newenv.push_back(NULL);
+
+    dup2(stdin_fd[0], STDIN_FILENO);
+    if (stdin_fd[0] != STDIN_FILENO) {
+      close(stdin_fd[0]);
+    }
+    dup2(stdout_fd[1], STDOUT_FILENO);
+    if (stdout_fd[1] != STDOUT_FILENO) {
+      close(stdout_fd[1]);
+    }
+    dup2(stderr_fd[1], STDERR_FILENO);
+    if (stderr_fd[1] != STDERR_FILENO) {
+      close(stderr_fd[1]);
+    }
+
+    if (execve(config_proto.filename().c_str(),
+               const_cast<char* const*>(&newargv[0]),
+               const_cast<char* const*>(&newenv[0]))) {
+      // Failed to Exec. Report through stderr, then die quickly and without
+      // side effects.
+      static const char kExecveMessage[] = "Child Unable to execve!\n";
+      write(STDERR_FILENO, kExecveMessage, sizeof(kExecveMessage) - 1);
+      while (1) {
+        _exit(1);
+      }
+    }
+  }
+  close(stdin_fd[0]);
+  close(stdout_fd[1]);
+  close(stderr_fd[1]);
+  {
+    std::unique_lock<std::mutex> write_lock(write_mutex_);
+    write_stream_.reset(new proto2::io::FileOutputStream(stdin_fd[1]));
+  }
+  {
+    std::unique_lock<std::mutex> read_lock(read_mutex_);
+    read_stream_.reset(new proto2::io::FileInputStream(stdout_fd[0]));
+  }
+  {
+    std::unique_lock<std::mutex> error_lock(error_mutex_);
+    error_stream_.reset(new proto2::io::FileInputStream(stderr_fd[0]));
+  }
+  pid_lock.unlock();
+  child_spawned_.notify_all();
+}
+
+void SubprocessDelegator::KillChildProcess() {
+  std::unique_lock<std::mutex> pid_lock(child_pid_mutex_);
+  // If there is no child, or if we are in the process of shutting down,
+  // do nothing.
+  if (child_pid_ <= 0) {
+    return;
+  }
+
+  kill(child_pid_, SIGTERM);
+  // Give it time to end gracefully, then make sure it is really dead.
+  std::this_thread::sleep_for(std::chrono::seconds(5));
+  kill(child_pid_, SIGKILL);
+  int child_status;
+  // Don't let it become a zombie.
+  waitpid(child_pid_, &child_status, 0);
```

Comment:
I am not sure the best course of action. On the one hand waiting
indefinitely for the child to die would block the parent. But if we do
something else then it might spawn an infinite number of children which all
end up zombies.

The current code waits a reasonable time (2 sec) before giving up on
killing the child FWIW:
https://github.com/google/grr/blob/master/client/nanny/windows_nanny.cc#L165


On 6 February 2015 at 13:38, bgalehouse <notifications@github.com> wrote:

> In client/minicomm/subprocess_delegator.cc
> <https://github.com/google/grr/pull/110#discussion_r24237846>:
>
> > +
> > +void SubprocessDelegator::KillChildProcess() {
> > +  std::unique_lock<std::mutex> pid_lock(child_pid_mutex_);
> > +  // If there is no child, or if we are in the process of shutting down,
> > +  // do nothing.
> > +  if (child_pid_ <= 0) {
> > +    return;
> > +  }
> > +
> > +  kill(child_pid_, SIGTERM);
> > +  // Give it time to end gracefully, then make sure it is really dead.
> > +  std::this_thread::sleep_for(std::chrono::seconds(5));
> > +  kill(child_pid_, SIGKILL);
> > +  int child_status;
> > +  // Don't let it become a zombie.
> > +  waitpid(child_pid_, &child_status, 0);
>
> Good idea. When I wrote this, my thought was that we can count on SIGKILL
> to finish the job reasonably quickly. However, I've since discovered that
> if the child is waiting on a system call, we might have to wait for that to
> finish. If we do a WNOHANG/sleep loop here and it times out, it would be
> good to prevent the child from spending an unbounded time as a zombie. Any
> suggestions on how to do this? I see two options:
>
> 1) Keep a list of undead children and regularly try to reap them. Seems a
> bit ugly, opens the possibility of complaining if the list becomes large
> (probably indicates we are trying to do something buggy)
>
> 2) disable SIGCHLD. Downside is that we no longer can count on child_pid
> existing. Perhaps it could even be reused on a sufficiently crazy busy
> system.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/grr/pull/110/files#r24237846>.
>